File: java/base/src/main/vsdk/toolkit/gui/tangibleInterfaces/TangibleInterfaceNetworkClient.java
package vsdk.toolkit.gui.tangibleInterfaces;

import java.net.URI;
import java.net.http.HttpClient;
import java.net.http.WebSocket;
import java.util.ArrayList;
import java.util.List;
import java.util.concurrent.CompletionStage;
import java.util.regex.Matcher;
import java.util.regex.Pattern;

import vsdk.toolkit.common.linealAlgebra.Quaterniond;
import vsdk.toolkit.common.linealAlgebra.Vector3Dd;

/**
This class connects to a `TangibleInterfaceMarkersDetectorServer` websocket
service (see `cpp/testsuite/Tools/TangibleInterfaceMarkersDetectorServer`),
listens for the JSON array of marker group poses it streams, and notifies
subscribed `TangibleInterfaceListener`s with a `TangibleInterfaceEvent` for
each reported marker group.

Each streamed JSON entry has the form:

<pre>
{"label":"rayCube","position":[x,y,z],"quaternion":[w,x,y,z]}
</pre>

If the service cannot be reached, this is reported on the console and the
client simply stops, without throwing an exception.
*/
public class TangibleInterfaceNetworkClient implements Runnable {

    private static final Pattern GROUP_PATTERN = Pattern.compile("\\{[^{}]*\\}");
    private static final Pattern LABEL_PATTERN = Pattern.compile("\"label\"\\s*:\\s*\"([^\"]*)\"");
    private static final Pattern POSITION_PATTERN = Pattern.compile("\"position\"\\s*:\\s*\\[([^\\]]*)\\]");
    private static final Pattern QUATERNION_PATTERN = Pattern.compile("\"quaternion\"\\s*:\\s*\\[([^\\]]*)\\]");

    private final String serviceUrl;
    private final List<TangibleInterfaceListener> listeners;
    private final StringBuilder incomingMessage;
    private WebSocket webSocket;

    /**
    @param serviceUrl websocket URL of the tangible interface marker
        tracking service, for example `ws://localhost:8090/v1/values`
    */
    public TangibleInterfaceNetworkClient(String serviceUrl) {
        this.serviceUrl = serviceUrl;
        this.listeners = new ArrayList<TangibleInterfaceListener>();
        this.incomingMessage = new StringBuilder();
    }

    /**
    Subscribes a listener to receive `TangibleInterfaceEvent`s as they
    arrive from the tracking service.

    @param listener the listener to subscribe
    */
    public void addListener(TangibleInterfaceListener listener) {
        if ( listener == null ) {
            return;
        }

        synchronized ( listeners ) {
            listeners.add(listener);
        }
    }

    /**
    Unsubscribes a previously added listener.

    @param listener the listener to unsubscribe
    */
    public void removeListener(TangibleInterfaceListener listener) {
        synchronized ( listeners ) {
            listeners.remove(listener);
        }
    }

    /**
    Spawns a separate thread that connects to the tangible interface
    tracking service and listens for incoming pose updates. This method
    returns immediately to its caller; the connection attempt and listening
    loop run on the spawned thread. If the service is not reachable, a
    message is printed to the console and the spawned thread terminates
    without raising an exception.
    */
    public void run() {
        Thread connectionThread = new Thread(this::connectAndListen, "TangibleInterfaceNetworkClient");
        connectionThread.setDaemon(true);
        connectionThread.start();
    }

    /**
    Requests the websocket connection to be closed, if currently open.
    */
    public void disconnect() {
        if ( webSocket != null ) {
            webSocket.sendClose(WebSocket.NORMAL_CLOSURE, "Client disconnect");
        }
    }

    private void connectAndListen() {
        HttpClient httpClient = HttpClient.newHttpClient();

        try {
            webSocket = httpClient.newWebSocketBuilder()
                .buildAsync(URI.create(serviceUrl), new FrameListener())
                .get();
        }
        catch ( Exception e ) {
            System.out.println("Tangible interface server not found at " + serviceUrl + ": " + e.getMessage());
        }
    }

    private void notifyListeners(TangibleInterfaceEvent event) {
        List<TangibleInterfaceListener> snapshot;

        synchronized ( listeners ) {
            snapshot = new ArrayList<TangibleInterfaceListener>(listeners);
        }

        for ( TangibleInterfaceListener listener : snapshot ) {
            listener.tangibleInterfaceEventReceived(event);
        }
    }

    private void processMessage(String message) {
        Matcher groupMatcher = GROUP_PATTERN.matcher(message);

        while ( groupMatcher.find() ) {
            TangibleInterfaceEvent event = parseEvent(groupMatcher.group());
            if ( event != null ) {
                notifyListeners(event);
            }
        }
    }

    private TangibleInterfaceEvent parseEvent(String groupJson) {
        String id = extractString(LABEL_PATTERN, groupJson);
        double[] positionValues = extractNumbers(POSITION_PATTERN, groupJson, 3);
        double[] quaternionValues = extractNumbers(QUATERNION_PATTERN, groupJson, 4);

        if ( id == null || positionValues == null || quaternionValues == null ) {
            return null;
        }

        Vector3Dd position = new Vector3Dd(positionValues[0], positionValues[1], positionValues[2]);
        Vector3Dd direction = new Vector3Dd(quaternionValues[1], quaternionValues[2], quaternionValues[3]);
        Quaterniond rotation = new Quaterniond(direction, quaternionValues[0]);

        return new TangibleInterfaceEvent(id, position, rotation);
    }

    private String extractString(Pattern pattern, String text) {
        Matcher matcher = pattern.matcher(text);

        if ( !matcher.find() ) {
            return null;
        }

        return matcher.group(1);
    }

    private double[] extractNumbers(Pattern pattern, String text, int count) {
        Matcher matcher = pattern.matcher(text);

        if ( !matcher.find() ) {
            return null;
        }

        String[] tokens = matcher.group(1).split(",");
        if ( tokens.length != count ) {
            return null;
        }

        double[] values = new double[count];
        for ( int i = 0; i < count; i++ ) {
            values[i] = Double.parseDouble(tokens[i].trim());
        }

        return values;
    }

    private class FrameListener implements WebSocket.Listener {
        public CompletionStage<?> onText(WebSocket webSocket, CharSequence data, boolean last) {
            incomingMessage.append(data);

            if ( last ) {
                String message = incomingMessage.toString();
                incomingMessage.setLength(0);
                processMessage(message);
            }

            webSocket.request(1);
            return null;
        }

        public void onError(WebSocket webSocket, Throwable error) {
            System.out.println("Tangible interface connection error on " + serviceUrl + ": " + error.getMessage());
        }
    }
}

File: cpp/base/src/main/vsdk/toolkit/gui/tangibleInterfaces/TangibleInterfaceNetworkClient.cpp
#include <cstdio>
#include <cstdlib>
#include <cstring>

#include "java/net/WebSocket.h"
#include "java/util/ArrayList.txx"
#include "vsdk/toolkit/gui/tangibleInterfaces/TangibleInterfaceListener.h"
#include "vsdk/toolkit/gui/tangibleInterfaces/TangibleInterfaceNetworkClient.h"

static int findText(const java::String& text, const char* needle, int fromIndex = 0)
{
    const char* base = text.c_str();
    const char* p = std::strstr(base + fromIndex, needle);
    if ( p == 0 ) {
        return java::String::npos;
    }
    return static_cast<int>(p - base);
}

TangibleInterfaceNetworkClient::TangibleInterfaceNetworkClient(const java::String& serviceUrl)
    : serviceUrl(serviceUrl),
      threadStarted(false),
      stopRequested(false),
      webSocket(0)
{
    pthread_mutex_init(&listenersMutex, 0);
}

TangibleInterfaceNetworkClient::~TangibleInterfaceNetworkClient()
{
    disconnect();
    pthread_mutex_destroy(&listenersMutex);
}

void TangibleInterfaceNetworkClient::addListener(TangibleInterfaceListener* listener)
{
    if ( listener == 0 ) {
        return;
    }

    pthread_mutex_lock(&listenersMutex);
    listeners.add(listener);
    pthread_mutex_unlock(&listenersMutex);
}

void TangibleInterfaceNetworkClient::removeListener(TangibleInterfaceListener* listener)
{
    pthread_mutex_lock(&listenersMutex);
    listeners.remove(listener);
    pthread_mutex_unlock(&listenersMutex);
}

void TangibleInterfaceNetworkClient::run()
{
    if ( threadStarted ) {
        return;
    }
    threadStarted = pthread_create(&connectionThread, 0, threadEntry, this) == 0;
}

void TangibleInterfaceNetworkClient::disconnect()
{
    stopRequested = true;
    if ( webSocket != 0 ) {
        webSocket->close();
    }
    if ( threadStarted && !pthread_equal(pthread_self(), connectionThread) ) {
        pthread_join(connectionThread, 0);
        threadStarted = false;
    }
    delete webSocket;
    webSocket = 0;
}

void* TangibleInterfaceNetworkClient::threadEntry(void* arg)
{
    TangibleInterfaceNetworkClient* client = static_cast<TangibleInterfaceNetworkClient*>(arg);
    if ( client != 0 ) {
        client->connectAndListen();
    }
    return 0;
}

void TangibleInterfaceNetworkClient::connectAndListen()
{
    webSocket = java::net::WebSocket::connect(serviceUrl);
    if ( webSocket == 0 ) {
        std::printf("Tangible interface server not found at %s\n", serviceUrl.c_str());
        return;
    }

    while ( !stopRequested ) {
        java::String payload;
        if ( !webSocket->readText(&payload) ) {
            break;
        }
        processMessage(payload);
    }
    if ( webSocket != 0 ) {
        webSocket->close();
    }
}

void TangibleInterfaceNetworkClient::notifyListeners(const TangibleInterfaceEvent& event)
{
    java::ArrayList<TangibleInterfaceListener*> snapshot;

    pthread_mutex_lock(&listenersMutex);
    for ( long i = 0; i < listeners.size(); i++ ) {
        snapshot.add(listeners.get(i));
    }
    pthread_mutex_unlock(&listenersMutex);

    for ( long i = 0; i < snapshot.size(); i++ ) {
        TangibleInterfaceListener* listener = snapshot.get(i);
        if ( listener != 0 ) {
            listener->tangibleInterfaceEventReceived(event);
        }
    }
}

void TangibleInterfaceNetworkClient::processMessage(const java::String& message)
{
    int len = message.length();
    for ( int i = 0; i < len; i++ ) {
        if ( message[i] != '{' ) {
            continue;
        }
        int start = i;
        while ( i < len && message[i] != '}' ) {
            i++;
        }
        if ( i >= len ) {
            break;
        }
        java::String groupJson = message.substr(start, i - start + 1);
        TangibleInterfaceEvent event;
        if ( parseEvent(groupJson, &event) ) {
            notifyListeners(event);
        }
    }
}

bool TangibleInterfaceNetworkClient::parseEvent(const java::String& groupJson, TangibleInterfaceEvent* outEvent)
{
    java::String id;
    double positionValues[3];
    double quaternionValues[4];

    if ( !extractString(groupJson, "label", &id) ||
         !extractNumbers(groupJson, "position", positionValues, 3) ||
         !extractNumbers(groupJson, "quaternion", quaternionValues, 4) ) {
        return false;
    }

    Vector3Dd position(positionValues[0], positionValues[1], positionValues[2]);
    Vector3Dd direction(quaternionValues[1], quaternionValues[2], quaternionValues[3]);
    Quaterniond rotation(direction, quaternionValues[0]);
    *outEvent = TangibleInterfaceEvent(id, position, rotation);
    return true;
}

bool TangibleInterfaceNetworkClient::extractString(const java::String& text, const char* key, java::String* out)
{
    java::String needle = java::String("\"") + key + "\"";
    int p = findText(text, needle.c_str());
    if ( p == java::String::npos ) {
        return false;
    }
    p = text.find(':', p + needle.length());
    if ( p == java::String::npos ) {
        return false;
    }
    p = text.find('"', p);
    if ( p == java::String::npos ) {
        return false;
    }
    int e = text.find('"', p + 1);
    if ( e == java::String::npos ) {
        return false;
    }
    *out = text.substr(p + 1, e - p - 1);
    return true;
}

bool TangibleInterfaceNetworkClient::extractNumbers(const java::String& text, const char* key, double* values, int count)
{
    java::String needle = java::String("\"") + key + "\"";
    int p = findText(text, needle.c_str());
    if ( p == java::String::npos ) {
        return false;
    }
    p = text.find('[', p + needle.length());
    int e = text.find(']', p);
    if ( p == java::String::npos || e == java::String::npos ) {
        return false;
    }

    java::String payload = text.substr(p + 1, e - p - 1);
    int start = 0;
    for ( int i = 0; i < count; i++ ) {
        int comma = payload.find(',', start);
        java::String token = (i == count - 1) ?
            payload.substr(start) : payload.substr(start, comma - start);
        if ( i != count - 1 && comma == java::String::npos ) {
            return false;
        }
        values[i] = std::atof(token.c_str());
        start = comma + 1;
    }
    return true;
}
